klippy/extras/exclude_object.py

```diff
@@ -0,0 +1,263 @@
+# Exclude moves toward and inside objects
+#
+# Copyright (C) 2019  Eric Callahan <arksine.code@gmail.com>
+# Copyright (C) 2021  Troy Jacobson <troy.d.jacobson@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import logging
+import json
+
+class ExcludeObject:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.gcode_move = self.printer.load_object(config, 'gcode_move')
+        self.printer.register_event_handler('klippy:connect',
+                                        self._handle_connect)
+        self.printer.register_event_handler("sdcard:reset_file",
+                                            self._reset_file)
+        self.next_transform = None
+        self.objects = {}
+        self.excluded_objects = []
+        self.current_object = None
+        self.in_excluded_region = False
+        self.extruder_idx = 0
+        self.extrusion_offsets = [[0., 0., 0., 0.]]
+        self.last_position = [0., 0., 0., 0.]
+        self.last_position_extruded = [0., 0., 0., 0.]
+        self.last_position_excluded = [0., 0., 0., 0.]
+        self.gcode.register_command(
+            'START_CURRENT_OBJECT', self.cmd_START_CURRENT_OBJECT,
+            desc=self.cmd_START_CURRENT_OBJECT_help)
+        self.gcode.register_command(
+            'END_CURRENT_OBJECT', self.cmd_END_CURRENT_OBJECT,
+            desc=self.cmd_END_CURRENT_OBJECT_help)
+        self.gcode.register_command(
+            'EXCLUDE_OBJECT', self.cmd_EXCLUDE_OBJECT,
+            desc=self.cmd_EXCLUDE_OBJECT_help)
+        self.gcode.register_command(
+            'EXCLUDE_OBJECT_RESET', self.cmd_EXCLUDE_OBJECT_RESET,
+            desc=self.cmd_EXCLUDE_OBJECT_RESET_help)
+        self.gcode.register_command(
+            'DEFINE_OBJECT', self.cmd_DEFINE_OBJECT,
+            desc=self.cmd_DEFINE_OBJECT_help)
+        self.gcode.register_command(
+            'LIST_OBJECTS', self.cmd_LIST_OBJECTS,
+            desc=self.cmd_LIST_OBJECTS_help)
+        self.gcode.register_command(
+            'LIST_EXCLUDED_OBJECTS', self.cmd_LIST_EXCLUDED_OBJECTS,
+            desc=self.cmd_LIST_EXCLUDED_OBJECTS_help)
+    def _setup_transform(self):
+        if not self.next_transform:
```

Comment:
It's preferable to use `if self.next_transform is None` when checking an external object because the external object could override the truth check.